### PR TITLE
Add nextjs + Auth0 starter

### PR DESCRIPTION
**Pull request by fstien**
Example app based on the nextjs starter project, with Auth0 as an authentication provider. 

Used the following Auth0 example: https://github.com/auth0-samples/auth0-golang-web-app/tree/master . I'm sure there are other ways of integrating with Auth0, such as with the nextjs-auth0 library. I wasn't able to get that to work though.

### CODE REVIEW

File: nextjs-auth0-starter/backend/auth/encore.gen.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Code generated by encore. DO NOT EDIT.
```
Comment (eandre) [inline nextjs-auth0-starter/backend/auth/encore.gen.go:1]: Code generated by Encore shouldn't be committed; it is generated on-demand

File: nextjs-auth0-starter/backend/auth/encore.gen.cue [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// Code generated by encore. DO NOT EDIT.
```
Comment (eandre) [inline nextjs-auth0-starter/backend/auth/encore.gen.cue:1]: Code generated by Encore shouldn't be committed; it is generated on-demand

File: nextjs-auth0-starter/backend/user/encore.gen.go [hunk 3]
```diff
@@ -0,0 +1,7 @@
+// Code generated by encore. DO NOT EDIT.
```
Comment (eandre) [inline nextjs-auth0-starter/backend/user/encore.gen.go:1]: Same here

**Review by eandre — COMMENTED** [review]
This looks great, thank you! A few minor nitpicks.

**Comment by fstien** [comment]
Thanks @eandre , I've removed all three generated files ; https://github.com/encoredev/examples/pull/50/commits/ed62e8da8557f7e5003b2d4975cbc7da325a23c3 .

**Comment by eandre** [comment]
Thanks @fstien, looks great!

Out of curiosity, do you recall what the issues were with using the nextjs-auth0 package? Just curious if there's some missing Encore functionality that would make it easier to integrate with it.

File: nextjs-auth0-starter/frontend/app/lib/client.ts [hunk 4]
```diff
@@ -0,0 +1,576 @@
...
+ */
+export type BaseURL = string
+
+export const Local: BaseURL = "http://localhost:4000"
+
+/**
+ * Environment returns a BaseURL for calling the cloud environment with the given name.
+ */
+export function Environment(name: string): BaseURL {
+    return `https://${name}-app-1234.encr.app`
```
Comment (simon-johansson) [inline nextjs-auth0-starter/frontend/app/lib/client.ts:19]:
```suggestion
    return `https://${name}-{{ENCORE_APP_ID}}.encr.app`
```

File: nextjs-auth0-starter/frontend/app/lib/client.ts [hunk 5]
```diff
@@ -0,0 +1,576 @@
...
+    readonly fetcher: Fetcher
+    readonly headers: Record<string, string>
+    readonly requestInit: Omit<RequestInit, "headers"> & { headers?: Record<string, string> }
+    readonly authGenerator?: AuthDataGenerator
+
+    constructor(baseURL: string, options: ClientOptions) {
+        this.baseURL = baseURL
+        this.headers = {
+            "Content-Type": "application/json",
+            "User-Agent":   "app-1234-Generated-TS-Client (Encore/v1.27.6)",
```
Comment (simon-johansson) [inline nextjs-auth0-starter/frontend/app/lib/client.ts:223]:
```suggestion
            "User-Agent":   "{{ENCORE_APP_ID}}-Generated-TS-Client (Encore/v1.27.6)",
```

**Review by simon-johansson — COMMENTED** [review]
Thanks a lot for contributing a starter! 👍

I made two small suggestions. Right now your app name is hard coded into the request client which would require the user to run `npm run gen:local` before running the frontend. 

Looks great otherwise, let's merge it after that has been fixed.

**Review by simon-johansson — APPROVED** [review]

**Comment by fstien** [comment]
@eandre I tried to use nextjs-auth0 a few months back, I can't remember all the details. All I can say is there was a lot of magic going on and things to configure (env variables, the auth0 console). I spent a lot of time messing with the config to get the basic auth flow to work with the redirect to the login page. At the end it was very flaky and broke for unclear reasons. 

In my experience auth0 does not have very good/well documented libraries. In this PR, there was something I had to figure out about the auth0 internals that was not obvious and not described clearly in the tutorial: https://auth0.com/docs/quickstart/webapp/golang/01-login . (I cannot remember what it was...)

I don't think this has anything to do with encore as all the auth logic happened in the nextJS project. The encore project was only implementing the auth token validation. 

Another aspect here is user management on the backend. With the next-auth0 approach it wasn't clear to me how to create a user row on the backend, which only sees auth tokens (which represent a claim on an auth0_user_id). I explored creating a table that maps auth0_user_id to an internal user_id which is read from and populated in the token validation logic. That didn't feel right, it adds latency and coupling to auth0 ( given the amount of Auth0 logic ).

With the approach in this PR, i can create a user in the callback of the signup flow (when the user gets redirected back to my nextjs frontend). The auth token validation remains stateless. 

Would be keen to hear about your experience with auth0 in production for the encore.dev console ? :)

**Pull request merged**